src/dmd/must_use.d

```diff
@@ -0,0 +1,203 @@
+module dmd.must_use;
+
+import dmd.dscope;
+import dmd.dsymbol;
+import dmd.expression;
+
+/**
+ * Check whether discarding an expression would violate the requirements of
+ * @mustUse. If so, emit an error.
+ *
+ * Params:
+ *   e = the expression to check
+ *   sc = scope in which `e` was semantically analyzed
+ *
+ * Returns: true on error, false on success.
+ */
+bool checkMustUse(Expression e, Scope* sc)
+{
+    import dmd.id : Id;
+
+    if (!isAssignment(e) && !isIncrementOrDecrement(e))
+    {
+        assert(e.type);
+        auto sym = e.type.toDsymbol(sc);
+        auto sd = sym ? sym.isStructDeclaration() : null;
+        // isStructDeclaration returns non-null for both structs and unions
+        if (sd && hasMustUseAttribute(sd, sc))
+        {
+            e.error("ignored value of `@%s` type `%s`; prepend a `cast(void)` if intentional",
+                Id.udaMustUse.toChars(), e.type.toChars());
+            return true;
+        }
+    }
+    return false;
+}
+
+/**
+ * Called from a symbol's semantic to check for reserved usage of @mustUse.
+ *
+ * If such usage is found, emits an errror.
+ *
+ * Params:
+ *   sym = symbol to check
+ */
+void checkMustUseReserved(Dsymbol sym)
+{
+    import dmd.id : Id;
+
+    if (sym.userAttribDecl is null || sym.userAttribDecl.atts is null)
+        return;
+
+    // Can't use foreachUda (and by extension hasMustUseAttribute) while
+    // semantic analysis of `sym` is still in progress
+    // TODO: factor out common code from this function and checkGNUABITag
+    foreach (exp; *sym.userAttribDecl.atts)
+    {
+        if (isMustUseAttribute(exp))
+        {
+            if (sym.isFuncDeclaration())
+            {
+                sym.error("`@%s` on functions is reserved for future use",
+                    Id.udaMustUse.toChars());
+                sym.errors = true;
+            }
+            else if (sym.isClassDeclaration() || sym.isInterfaceDeclaration() || sym.isEnumDeclaration())
```

Comment:
```suggestion
            else if (sym.isClassDeclaration() || sym.isEnumDeclaration())
```

The check is redundant as `InterfaceDeclaration` inherits from `ClassDeclaration`